core/engine/src/object/builtins/jsarraybuffer.rs

```diff
@@ -195,6 +197,78 @@ impl JsArrayBuffer {
                     .into()
             })
     }
+
+    /// Get an immutable reference to the [`JsArrayBuffer`]'s data.
+    ///
+    /// Returns `None` if detached.
+    ///
+    /// ```
+    /// # use boa_engine::{
+    /// # object::builtins::JsArrayBuffer,
+    /// # Context, JsResult, JsValue
+    /// # };
+    /// # fn main() -> JsResult<()> {
+    /// # // Initialize context
+    /// # let context = &mut Context::default();
+    /// // Create a buffer from a chunk of data
+    /// let data_block: Vec<u8> = (0..5).collect();
+    /// let array_buffer = JsArrayBuffer::from_byte_block(data_block, context)?;
+    ///
+    /// // Get a reference to the data.
+    /// let internal_buffer = array_buffer.data();
+    ///
+    /// assert_eq!(internal_buffer.as_deref(), Some((0..5).collect::<Vec<u8>>().as_slice()));
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[inline]
+    #[must_use]
+    pub fn data(&self) -> Option<GcRef<'_, [u8]>> {
+        GcRef::try_map(
+            self.inner
+                .downcast_ref::<ArrayBuffer>()
+                .expect("inner must be an ArrayBuffer"),
```

Comment:
Can we avoid the panic? It seems preferable to just return `None` in that case, since the panic shouldn't be reachable anyways.

Reply:
I thought of doing that, but if the invariant is broken then this does not panic, the bug will propagate somewhere else.

Though I'm fine with changing it if you want :)

A thought: It would be interesting if we had something like `JsObject<T = OridinaryObject>` storing that as the inner, that way we can avoid any checks and can access then data without `.downcast_ref()`... At that point we could make `JsArrayBuffer` a type alias to `JsObject<ArrayBuffer>` and just implement the methods on that type `impl JsObject<ArrayBuffer> { ... }`.

Reply:
> I thought of doing that, but if the invariant is broken then this does not panic, the bug will propagate somewhere else.

We could just add a `debug_assert!` for good measure. IMHO, the advantages of not having another panic in our code outweigh a theoretical bug that we haven't encountered yet.

> A thought: It would be interesting if we had something like `JsObject<T = OridinaryObject>` storing that as the inner, that way we can avoid any checks and can access then data without `.downcast_ref()`... At that point we could make `JsArrayBuffer` a type alias to `JsObject<ArrayBuffer>` and just implement the methods on that type `impl JsObject<ArrayBuffer> { ... }`.

Yep, I'm planning to experiment a bit with that, to see how much we can leverage the new object representation.